### ISSUE TITLE
Strict list pattern matching, reportMatchNotExhaustive false positive

### ISSUE THREAD

**Issue by kwshi** [microsoft/pyright#6843]
**Describe the bug**
When pattern-matching on a list, the cases `[]` and `[a, *b]` account for all possibilities (at least, I believe this is true in Python; it certainly is true in all the FP languages that pioneered structural pattern-matching, where "empty list" and "cons" are by definition the two cases a list can fall into). Currently, Pyright fails to recognize this, and incorrectly raises a "Cases within match statement do not exhaustively handle all values" error in strict mode.

**Code or Screenshots**

```python
# pyright: strict

def stuff1(x: list[int]):
    match x:
        case []:
            print("empty")
        case [a, *b]:
            print(f"non-empty: {a} {b}")

def stuff2(x: list[int]):
    match x:
        case [*c]:
            print(f"any list, {c}")
```

![image](https://github.com/microsoft/pyright/assets/3887189/60fd2171-5e75-4a82-a18f-8d1b538068c7)

`stuff1` is an example where the list pattern-matching is exhaustive, but Pyright is incorrectly flagging it as non-exhaustive. `stuff2` is only included for comparison, where Pyright correctly deduces that `[*c]` is fully general and doesn't raise an error; it's meant to show that Pyright isn't blindly treating all destructuring patterns as non-exhaustive, as it does with e.g. pattern guards (`case ... if ...`), but that it seems to be selectively recognizing certain destructuring patterns and not others.

**VS Code extension or command-line**
As language server via Neovim LSP, and also invoked directly from CLI. Version 1.1.343.

**Comment by erictraut** [comment]
Pyright is working as intended here. Pattern matching exhaustion is detected via type narrowing. In your example, the type of `list[int]` cannot be narrowed by the first pattern (`[]`) because there's no way in the Python type system to represent a "list that contains at least one element". You'll need to use a different strategy if you want pyright (or any other Python static type checker, for that matter) to detect exhaustion.
